Implemented Describe > Three Variable Summarise

Vitalis95 opened the pull request:
Still in progress
@rdstern , have a look at the progress made

rdstern commented:
@Vitalis95 that's looking very promising.  Here is the option for 2 variables when Numeric by Numeric

![image](https://github.com/user-attachments/assets/ed697e8f-2afb-46a4-b932-945b56e0bb22)

a) In the 3 variable case, with the first and second numeric could you add the Means and Sig Level as well and also the Swap and the misssing values option.
b) @lilyclements may want to comment on this point.  I suggest, when we have Means (both 2 variable and 3 Variable) we should change the word Means to Model.  The Model is the means when a variable is a factor, and it is the Slope (or Slopes) for a Variate.  So that's what we should be giving there?  
c) In 3 variable with ANOVA could there be another option, with now some radio buttons.  There will be more, but this one to start with is `Interaction`.  If chosen then it gives the ANOVA with the interaction term.  
d) And a third radio button could we have Compare as the label.  This would not have the option of means.  It would include sig level and the swap.  I forget the command, but it would give the ANOVA line with just the second variable, then the third and then the intereaction.  So one could find the decrease in the residual when we add the third variable (so the second in the x of the model, and then the extra when we also add the interaction term.

Vitalis95 commented:
> @Vitalis95 that's looking very promising. Here is the option for 2 variables when Numeric by Numeric
> 
> ![image](https://private-user-images.githubusercontent.com/11226469/350050422-ed697e8f-2afb-46a4-b932-945b56e0bb22.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzNDAxOTgsIm5iZiI6MTcyMjMzOTg5OCwicGF0aCI6Ii8xMTIyNjQ2OS8zNTAwNTA0MjItZWQ2OTdlOGYtMmFmYi00NmE0LWI5MzItOTQ1YjU2ZTBiYjIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzMwVDExNDQ1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3YmNiMDUxMGI5Nzc2NGQ2NGE2Zjg3NjIwNGIxOWQ0YTFmNDc3Mjc5MzM1Mjc4NTJmMzc0NGQ1MTNlMTk4MjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eawf9F9gMuPHT0q6bpp75_SIjOYYTSReCew868eZqpE)
> 
> a) In the 3 variable case, with the first and second numeric could you add the Means and Sig Level as well and also the Swap and the misssing values option. b) @lilyclements may want to comment on this point. I suggest, when we have Means (both 2 variable and 3 Variable) we should change the word Means to Model. The Model is the means when a variable is a factor, and it is the Slope (or Slopes) for a Variate. So that's what we should be giving there? c) In 3 variable with ANOVA could there be another option, with now some radio buttons. There will be more, but this one to start with is `Interaction`. If chosen then it gives the ANOVA with the interaction term. d) And a third radio button could we have Compare as the label. This would not have the option of means. It would include sig level and the swap. I forget the command, but it would give the ANOVA line with just the second variable, then the third and then the intereaction. So one could find the decrease in the residual when we add the third variable (so the second in the x of the model, and then the extra when we also add the interaction term.

@lilyclements , what is your comment on this? The interaction term is not in the `$anova_tables2`  function

rdstern commented:
@lilyclements and @Vitalis95 I was delaying on this dialog.  But it has now become urgen, because the merged version of this dialog, doesn't do the ANOVA at all.  It gives a bug.  The 2-way ANOVA is working on this dialog, so let's merge it as soon as we can.
If @lilyclements has time this week to improve the function, then let's do this too?  If so, then here are the 2 items it would be good to add for the 2-variable ANOVA options.  Lily may have more?
a) Add a `Total` checkbox, default unchecked.
b) Change the Means option to Means/Model.  Lily, currently it gives the means whatever the x.  Here is an example:

![image](https://github.com/user-attachments/assets/e82560df-a5e2-4840-95bf-28b1288aeb77)

When x is numeric it should give the intercept and slope.

In this example the means are even more ridiculous if you swap the variables.

lilyclements commented:
@rdstern thanks for this - I've fixed (a) and (b) in this PR #9152

rdstern commented:
@Vitalis95 @lilyclements has now added trhe option for the Total and changed the Means into Means/Model.  That's great.
On layout (you may have better ideas.  I wonder about 2 rows of these options now
Checkbox Total instead of Means and then Sig Level as now.
Then the Means Checkbox on the line below, by itself, and now written as Means/Model.  I assume that it will give either the means or the model, automatically, depending on whether the x is numeric or a factor.

Vitalis95 commented:
@rdstern  about **c)**, you had suggested earlier the the `Swarp First/Second variables` checkbox be visible, when checked it gives `ANOVA`, when unchecked it gives `summaries` that is when we have  `categorical by numeric`

rdstern commented:
@Vitalis95 that's right.  Did I get that wrong in my checking?  That would be a relief.  So now, does categorical by numeric still give summary tables?  And when you swap it gives ANOVA?  If so, that's great and please disregard c).

Vitalis95 commented:
> @Vitalis95 @lilyclements has now added trhe option for the Total and changed the Means into Means/Model. That's great. On layout (you may have better ideas. I wonder about 2 rows of these options now Checkbox Total instead of Means and then Sig Level as now. Then the Means Checkbox on the line below, by itself, and now written as Means/Model. I assume that it will give either the means or the model, automatically, depending on whether the x is numeric or a factor.

@lilyclements , following this comment, could the` total` be passed like the `sign_level and mean` arguments, such that when the Total checkbox is checked it produces the totals

rdstern commented:
@Vitalis95 that's exactly what I was hoping for with the Total line.

Vitalis95 commented:
@lilyclements , I pulled the changes and now there is an error, 



![image](https://github.com/user-attachments/assets/1c31b48e-539f-4ff3-a603-9c046ce4cd62)

lilyclements commented:
@Vitalis95 thanks for this - yes, this is a great point. I forgot to update `anova_tables2` in the instat_object after updating in the data_object file. This is done [here](https://github.com/Vitalis95/R-Instat/pull/281/files) and ready for you to merge into this branch if you are happy with it.

rdstern commented:
@Vitalis95 and @lilyclements this is looking great!
a) Vitalis I remember Lily mentioned that the default was to give the total, but you could change that, in her function.  Pleasec could you do that.  Currently it gives the total whatever I do.  The default is not to give the total, because that's what R does.
b) Please could you move Sig Level up, so it is after the total.  (They are both concerned with the anova table, so they come above the Means/Model.
c) @lilyclements everything is there now, but can the output be improved?  I hope that's not too time consuming.  Could Francis help if needed?  Here is an example:

![image](https://github.com/user-attachments/assets/db918b07-5edc-4c27-98cd-61299285fa2a)

1. It isn't a Means table of yield.  I presume we could say just Model?
2. Could there be blanks instead of NA?
3. Could the F probability be formatted better?
4. And do we need to know it is a tibble and have all the [[1]] and list the type of each variable?

Vitalis95 commented:
@rdstern , have a look at it

rdstern commented:
@Vitalis95 looks really good now.  I'll just wait to see if Lily makes any changes?

lilyclements reviewed (dismissed):
@Vitalis95 looks good to me too!

lilyclements commented:
@Vitalis95 @rdstern apologies didn't mean to approve. I will make some changes to the R code

lilyclements commented:
@rdstern I made the changes you requested. You can now test it

rdstern reviewed (commented):
@lilyclements you now give the means twice, and not the ANOVA table!
If possible could you, or @Vitalis95 also improve the spacing of the controls, when there are both numeric variables

lilyclements commented:
@rdstern so I did, apologies. I've fixed that now (hopefully!). Very sorry for that

rdstern reviewed (commented):
@lilyclements that's looking much better now.  
![image](https://github.com/user-attachments/assets/c58ec9d4-7b13-4f95-8803-60819040ef40)

a) Great the F-prob like that now.
b) Could you swap the order back, so the ANOVA comes before the means or the model?  I see the ANOVA as the basic result, optionally followed by the model or means when needed.
c) Can you get rid of the extra tibble details, and the [[1]], etc?  If that's too big to do now, then we could approve with these, and then improve the presentation when we add the 3-variable stuff in a new pull request?

lilyclements reviewed (commented):
@Vitalis95 I've updated `map` to be `walk` for two classes/functions? 

Are these classes functions only used for `anova_table2` bits?

`walk` means that it will not print output, so will not give the list output!

@Vitalis95 are you happy with these changes? (Make sure you pull changes too next time you work on this - I've made changes now sorry!)

@rdstern does this now work for you how you would like it?

rdstern reviewed (dismissed):
@lilyclements I was not sure what has changed.  There is still the tibble-like output.

I am approving, because it is now so much better than the vurrent merged one.  We will add another pull request for the 3 variable case, and that also gives another opportunity to improve the formatting of this output.

@N-thony over to you to check, and hopefully merge.

Vitalis95 commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -1094,19 +1219,32 @@ Public Class dlgDescribeTwoVariable
+                '    '    Else
+                '    '        clsCombineAnova2Function.AddParameter("y", ucrReceiverThreeVariableThirdVariable.GetVariableNames(True), iPosition:=2, bIncludeArgumentName:=False)
+
+                '    '    End If
```
done

Vitalis95 commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -625,20 +643,25 @@ Public Class dlgDescribeTwoVariable
+            '    ucrChkOmitMissing.Visible = IsFactorByNumericByNumeric() OrElse IsNumericByNumericByFactor()
+            'Else
+            '    ucrChkOmitMissing.Visible = False
+            '    cmdMissingOptions.Visible = False
```
done

Vitalis95 commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -625,20 +643,25 @@ Public Class dlgDescribeTwoVariable
-            ucrInputMarginName.Visible = ucrChkDisplayMargins.Checked AndAlso IsFactorByFactorByFactor()
+            ucrChkSwapXYVar.Visible = IsNumericByNumericByNumeric() OrElse IsNumericByNumericByFactor()
+            'ucrChkDisplayMargins.Visible = IsFactorByFactorByFactor()
+            'ucrInputMarginName.Visible = ucrChkDisplayMargins.Checked AndAlso IsFactorByFactorByFactor()
```
done

Vitalis95 commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -761,45 +786,93 @@ Public Class dlgDescribeTwoVariable
+                '                                  strRObjectTypeLabelToAssignTo:=RObjectTypeLabel.Table,
+                '                                      strRObjectFormatToAssignTo:=RObjectFormat.Html,
+                '                                      strRDataFrameNameToAddObjectTo:=ucrSelectorDescribeTwoVar.strCurrentDataFrame,
+                '                                        strObjectName:="last_table")
```
done

Vitalis95 commented:
@N-thony , have a look at it

rdstern reviewed (approved):
@N-thony I am approving, as this is an important advance.  It will be good for it to be merged.
@Vitalis95 please can you make a new pull request for the Three Variable part.  That will be even more exciting.  And remind @lilyclements about the R bit, for parallel and full, as well as making the result neater for all the ANOVA.
